### PR TITLE
SC development best practices

### PR DESCRIPTION
**Pull request by annenkov**
## Purpose

A section on best practices of SC development.

### CODE REVIEW

**Comment by dg-concordium** [comment]
@annenkov Will there be more best practices coming? You have an index file so I thought maybe there will be more in future. Otherwise, I would probably get rid of the index file for best practices and just insert the file into the TOC.

**Review by dg-concordium — APPROVED** [review]

**Comment by annenkov** [comment]
> @annenkov Will there be more best practices coming? You have an index file so I thought maybe there will be more in future. Otherwise, I would probably get rid of the index file for best practices and just insert the file into the TOC.

@dg-concordium Yes, I thought of it as a folder where we could put additional material, e.g. detailed descriptions of certain security issues, testing and verification techniques, audit best practices, etc.

**Review by abizjak — APPROVED** [review]

File: source/mainnet/net/index.rst [hunk 1]
```diff
@@ -72,6 +72,7 @@ index
    :caption: Smart Contracts v1
 
    Introduction <../smart-contracts/general/introduction>
+   Best practices <../smart-contracts/best-practices/index>
```
Comment (Bargsteen) [inline source/mainnet/net/index.rst:75]: I don't think this should be right after the introduction. Perhaps after the on-chain guides but before Ethereum?

File: source/mainnet/smart-contracts/best-practices/index.rst [hunk 2]
```diff
@@ -0,0 +1,12 @@
+. _sc-best-practices:
+
+==============
+Best practices
+==============
+
+Here, you can find best practices for smart contract development, audit, information about common pitfalls and security vulnerabilities, and how to avoid them.
```
Comment (Bargsteen) [inline source/mainnet/smart-contracts/best-practices/index.rst:7]: I think we should remove this page for now. Then we can add it once we have more than one page of best practices. I don't see much benefit in having it now, and it just adds an extra click for the user.

File: source/mainnet/smart-contracts/best-practices/development.rst [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+
+Mindset
+========
+
+Smart contract development involves many risks that do not show up in, for example, web programming:
+
+- the cost of mistakes is very high;
+- possibilities for fixing bugs are limited;
+- the area is evolving constantly, with new vulnerabilities being discovered regularly;
+- malicious parties deliberately try to break your smart contracts and steal the money.
```
Comment (Bargsteen) [inline source/mainnet/smart-contracts/best-practices/development.rst:18]: I don't know if "steal the money" is too specific. There are types of hacks that don't involve money directly.

File: source/mainnet/smart-contracts/best-practices/development.rst [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+You can think about smart contracts as mission-critical software, or software for embedded devices rather than a web application.
+
+To minimize the exposure of your smart contract to possible attacks consider the following.
+
+- Determine the “bare minimum” that should be on-chain: use smart contracts only for the part that requires decentralization.
+- Keep your contracts simple and minimalistic: stick to strictly necessary functionality only.
+  The more complexity you have in your contract, the larger the attack surface.
+- Be ready if things go wrong after deployment:
+
+  - provide “pause” functionality;
```
Comment (Bargsteen) [inline source/mainnet/smart-contracts/best-practices/development.rst:30]: We should also mention the native upgrades here.

File: source/mainnet/smart-contracts/best-practices/development.rst [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+This section provides recommendations for developing smart contracts in Rust.
+
+
+.. _best-practices-code-structure:
+
+Recommended structure
+---------------------
+
+- Use ``cargo init`` with an appropriate template to start a new project.
+- Build contract logic as the state struct implementation.
```
Comment (Bargsteen) [inline source/mainnet/smart-contracts/best-practices/development.rst:76]: Is this what we recommend?

File: source/mainnet/smart-contracts/best-practices/development.rst [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+.. _best-practices-external-calls:
+
+External Calls
+--------------
+
+Every external call should be treated as a potential security risk.
+Calling another contract gives control to potentially malicious code that could make arbitrary calls to any other contract, including your own contract.
+Calls to your contract might change its state through entrypoints that permit updating the state, see :ref:`best-practices-reentrancy`.
+Moreover, you should not make any assumptions about energy consumption, or expect that the execution succeeds.
+You contract should be able to correctly handle situations when the call to an external contract fails.
```
Comment (Bargsteen) [inline source/mainnet/smart-contracts/best-practices/development.rst:125]:
```suggestion
Your contract should be able to correctly handle situations when the call to an external contract fails.
```

File: source/mainnet/smart-contracts/best-practices/development.rst [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+
+General recommendations
+^^^^^^^^^^^^^^^^^^^^^^^
+
+- *Avoid complex interactions*. Avoid splitting the on-chain part of your dApp into several smart contracts unless it is strictly necessary.
+  For example, instead of using the *proxy* pattern, use :ref:`natively upgradable contracts <contract-instance-upgradeability>`.
```
Comment (Bargsteen) [inline source/mainnet/smart-contracts/best-practices/development.rst:131]: Please add a link or more description to "the proxy pattern". People might not be aware of this pattern.

File: source/mainnet/smart-contracts/best-practices/development.rst [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+^^^^^^^^^^
+
+The *reentrant behavior* is not specific to smart contracts: it is a well known issue in the context of concurrency.
+A procedure can be interrupted in the middle of its execution, run again in *another* execution context, and then continue execution from the interruption point.
+In case of smart contracts, each external call interrupts the execution and hands over control to unknown code.
+Do not treat external contract invocations as regular method calls.
+Instead, think of them as sending a message and temporarily pausing execution of your contract.
+The receiving side has full control of what to do next and can choose to call your contract again while it is still in the "paused" state waiting for the external call to be completed.
+Once the external call is completed, you might find your contract in a completely different state.
+See an :ref:`example <reentracny-unit-testing>`, based on the DAO contract vulnerability of how reentrancy can be discovered using unit testing.
```
Comment (Bargsteen) [inline source/mainnet/smart-contracts/best-practices/development.rst:150]: Please add a link or description to the DAO contract vulnerability.

File: source/mainnet/smart-contracts/best-practices/development.rst [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+
+- Avoid changing the state after an external call: use the *Checks-Effects-Interactions* pattern: validate data, update the contract state, make external calls.
+- If you need to perform some state changes after an external call use `invoke_contract_read_only <https://docs.rs/concordium-std/latest/concordium_std/trait.HasHost.html#method.invoke_contract_read_only>`_.
+  If the read-only invocation succeeds, it ensures that the state has not been changed after returning from the external call.
+  Using ``invoke_contract`` covers most of the cases where protection from reentrancy is required.
```
Comment (Bargsteen) [inline source/mainnet/smart-contracts/best-practices/development.rst:155]: Did you mean `invoke_contract_read_only` here ?

File: source/mainnet/smart-contracts/best-practices/development.rst [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+- Alternatively, consider using a *mutex*: a boolean flag that is set before making an external call, preventing all entrypoints from reentrancy. Reset after the call is complete.
```
Comment (Bargsteen) [inline source/mainnet/smart-contracts/best-practices/development.rst:156]: Perhaps we could include a few lines of code to illustrate this?

File: source/mainnet/smart-contracts/best-practices/development.rst [hunk 11]
```diff
@@ -0,0 +1,203 @@
...
+------------------
+
+- Write an outline of the smart contract functionality in the beginning of the file; if the contract implements some standards, mention it.
+- Document decisions/choices in the code.
+- Document entrypoints:
+
+  - What functionality the entrypoint implements?
+  - Who has access rights to call the entrypoint?
+  - When the call is rejected?
+  - What events are logged?
```
Comment (Bargsteen) [inline source/mainnet/smart-contracts/best-practices/development.rst:170]: and when?

File: source/mainnet/smart-contracts/best-practices/development.rst [hunk 12]
```diff
@@ -0,0 +1,203 @@
...
+
+Automated testing
+-----------------
+
+The Concordium standard library `concordium-std`_ offers several possibilities for testing the smart contract code.
+
+- :ref:`Unit testing <unit-test-contract>` used for testing particular cases where you define what is the valid output.
+- :ref:`Property-based testing <writing_property_based_tests>` is a variant of randomized testing that repeatedly checks a *property* with randomly generated input.
+
+Use the :ref:`smart contract specification <best-practices-specification>` to come up with cases and properties to test.
```
Comment (Bargsteen) [inline source/mainnet/smart-contracts/best-practices/development.rst:187]: It wasn't clear to me that this referred to a section in this page.

**Review by Bargsteen — CHANGES_REQUESTED** [review]
Looks good. Had a few comments 😊

File: source/mainnet/smart-contracts/best-practices/index.rst (hunk 2 above)
Comment (annenkov) [reply source/mainnet/smart-contracts/best-practices/index.rst:7]: I wanted to create a structure that could be filled-in with new content, and possibly write some short guidelines on how to contribute. So I could drive the process (and contribute myself), but also encourage contributions by others.
We can discuss how to organise the process on one of the stand-ups.

File: source/mainnet/smart-contracts/best-practices/development.rst (hunk 3 above)
Comment (annenkov) [reply source/mainnet/smart-contracts/best-practices/development.rst:18]: True, but stealing funds are one of the common goals (and probably the main motivation).
Maybe, "try to steal funds or break your contract in some other way"?

File: source/mainnet/smart-contracts/best-practices/development.rst (hunk 5 above)
Comment (annenkov) [reply source/mainnet/smart-contracts/best-practices/development.rst:76]: I've talked to @abizjak about this at some point, and my understanding was that this is a recommended structure. Most of our non-trivial examples follow this approach. But I might be wrong about this :) So, I'm open to other suggestions.

File: source/mainnet/smart-contracts/best-practices/development.rst (hunk 9 above)
Comment (annenkov) [reply source/mainnet/smart-contracts/best-practices/development.rst:155]: Yes! Good catch, thanks! ;)

File: source/mainnet/smart-contracts/best-practices/development.rst (hunk 12 above)
Comment (annenkov) [reply source/mainnet/smart-contracts/best-practices/development.rst:187]: I don't fully understand what you mean here. Do you mean there is another way of referencing the content on the same page? Could you please clarify?

File: source/mainnet/smart-contracts/best-practices/development.rst [hunk 13]
```diff
@@ -0,0 +1,203 @@
...
+Moreover, you should not make any assumptions about energy consumption, or expect that the execution succeeds.
+You contract should be able to correctly handle situations when the call to an external contract fails.
+
+General recommendations
+^^^^^^^^^^^^^^^^^^^^^^^
+
+- *Avoid complex interactions*. Avoid splitting the on-chain part of your dApp into several smart contracts unless it is strictly necessary.
+  For example, instead of using the *proxy* pattern, use :ref:`natively upgradable contracts <contract-instance-upgradeability>`.
+- *Think about the contract state*. Do not assume that the contract state stays the same after an external call (see :ref:`best-practices-reentrancy` for details).
+- *Protect from denial-of-service (DoS) attacks*. Use the *Pull over Push* pattern: avoid sending funds back (*Push*) to an unknown address as part of some complex operation in your smart contract.
```
[the remark this answers is not in the document]
Comment (annenkov) [reply source/mainnet/smart-contracts/best-practices/development.rst:133]: Yeah, I agree, there should be a whole section about DoS vulnerabilities :) I can mention these issues without much detail for now.

File: source/mainnet/smart-contracts/best-practices/development.rst (hunk 7 above)
Comment (annenkov) [reply source/mainnet/smart-contracts/best-practices/development.rst:131]: I'm not sure where to refer. We have an example, but it's in a pull request. Should I reference some Solidity proxy pattern then? Maybe we can just assume if people don't know, then they don't need to know about it anyways. :)

File: source/mainnet/smart-contracts/best-practices/development.rst (hunk 3 above)
Comment (Bargsteen) [reply source/mainnet/smart-contracts/best-practices/development.rst:18]: Yea, just phrase it as an example of things that can occur.

File: source/mainnet/smart-contracts/best-practices/development.rst (hunk 5 above)
Comment (Bargsteen) [reply source/mainnet/smart-contracts/best-practices/development.rst:76]: I mean it is fine and we do use it. I was just unaware that it was now the recommended approach ;)

File: source/mainnet/smart-contracts/best-practices/development.rst (hunk 7 above)
Comment (Bargsteen) [reply source/mainnet/smart-contracts/best-practices/development.rst:131]: I don't like referring to something that users perhaps might not know. We could link to a solidity pattern, mention that it makes everything more complex, and then say that on the superior Concordium blockchain, we can use native upgradability.

File: source/mainnet/smart-contracts/best-practices/development.rst (hunk 12 above)
Comment (Bargsteen) [reply source/mainnet/smart-contracts/best-practices/development.rst:187]: When I read it, I thought it referred to some other page or guideline and then I was briefly confused about it leading to a section on the same page.

I would just add a few more words, e.g.
```suggestion
Use the :ref:`smart contract specification <best-practices-specification>` guidelines from the section above to come up with cases and properties to test.
```

File: source/mainnet/smart-contracts/best-practices/development.rst [hunk 14]
```diff
@@ -0,0 +1,203 @@
...
+  It opens doors to DoS attacks since the contract you call might fail for various reasons, blocking your functionality from succeeding.
+  Instead, create a separate entrypoint allowing users, which could be smart contracts, to request funds back (*Pull*).
```
[the remark this answers is not in the document]
Comment (abizjak) [reply source/mainnet/smart-contracts/best-practices/development.rst:135]: Yes, I understand what you are referring to. I just think the way that it is explained now makes something that is nuanced, and different approaches have tradeoffs, into a one-sided recommendation.

I also don't understand how the "push pull" terminology is enlightening. To me it does not say much.

[the remark this answers is not in the document]
Comment (annenkov) [reply source/mainnet/smart-contracts/best-practices/development.rst:135]: > I just think the way that it is explained now makes something that is nuanced, and different approaches have tradeoffs, into a one-sided recommendation.

Fair enough, I'll reformulate it.

> I also don't understand how the "push pull" terminology is enlightening. To me it does not say much.

I agree, but this "pull-over-push" is something that is used often, it's not a "standard", but at least something people can recognise, so I think it makes sense to use the same terminology.

**Comment by annenkov** [comment]
@Bargsteen I've addressed the comments. The one remaining is whether we should keep the index page referring to one page or not. My point is that it creates some structure for further contributions, so we just add stuff without changing the structure. But I don't know when we're going to have more pages there :)

File: source/mainnet/smart-contracts/best-practices/development.rst [hunk 15]
```diff
@@ -0,0 +1,295 @@
...
+You can think about smart contracts as mission-critical software, or software for embedded devices rather than a web application.
+
+To minimize the exposure of your smart contract to possible attacks consider the following.
+
+- Determine the “bare minimum” that should be on-chain: use smart contracts only for the part that requires decentralization.
+- Keep your contracts simple and minimalistic: stick to strictly necessary functionality only.
+  The more complexity you have in your contract, the larger the attack surface.
+- Be ready if things go wrong after deployment:
+
+  - provide “pause” functionality;
```
Comment (Bargsteen) [inline source/mainnet/smart-contracts/best-practices/development.rst:30]: We mention pausing several times without really describing it. Could you add a note about it or link to some example?

File: source/mainnet/smart-contracts/best-practices/development.rst [hunk 16]
```diff
@@ -0,0 +1,295 @@
...
+        name = "doSomething",
+        parameter = "MyParameter",
+        mutable
+    )]
+    fn contract_do_something<S: HasStateApi>(
+        ctx: &impl HasReceiveContext,
+        host: &mut impl HasHost<State<S>, StateApiType = S>,
+    ) -> ReceiveResult<()> {
+        // Parse parameters
+        let param : MyParameter = ctx.parameter_cursor().get()?;
```
Comment (Bargsteen) [inline source/mainnet/smart-contracts/best-practices/development.rst:105]:
```suggestion
        let param: MyParameter = ctx.parameter_cursor().get()?;
```

File: source/mainnet/smart-contracts/best-practices/development.rst [hunk 17]
```diff
@@ -0,0 +1,295 @@
...
+External call failure
+^^^^^^^^^^^^^^^^^^^^^
+
+Sending funds back to an unknown contract address as part of some complex operation could block this operation from succeeding.
+The contract you call might fail for various reasons.
+If the contract call fails, the whole operation also fails.
+
+Consider splitting withdrawal of funds from the rest of the contract logic.
+You could create a separate entrypoint allowing users, which could be smart contracts, to request funds back.
+This pattern is called *Pull over Push*, where *Pull* corresponds to user explicitly requesting funds and *Push* to sending the funds back as part of some other operation.
```
Comment (Bargsteen) [inline source/mainnet/smart-contracts/best-practices/development.rst:136]:
```suggestion
This pattern is called *Pull over Push*, where *Pull* corresponds to the user explicitly requesting funds and *Push* to sending the funds back as part of some other operation.
```

File: source/mainnet/smart-contracts/best-practices/development.rst [hunk 18]
```diff
@@ -0,0 +1,295 @@
...
+
+Computing the maximum for each new bid requires iterating over the map.
+As the number of participants grows it becomes more expensive to compute the highest bid.
+Eventually, it might not fit into the block energy limit and bidding becomes blocked.
+See :ref:`contract-instance-operations` for more information.
+
+This situation is not necessarily an attack, it could occur naturally during the contract lifetime.
+In general, all computations requiring iteration with no clear bound could be an issue.
+For the auction, consider adding ``previous_bid: Amount`` to the state.
+Update ``previous_bid`` once a new bidder proposes a higher bid.
```
Comment (Bargsteen) [inline source/mainnet/smart-contracts/best-practices/development.rst:164]:
```suggestion
For the auction, consider adding ``highest_bid: Amount`` to the state.
Update ``highest_bid`` once a new bidder proposes a higher bid.
```

File: source/mainnet/smart-contracts/best-practices/development.rst [hunk 19]
```diff
@@ -0,0 +1,295 @@
...
+
+.. note::
+
+  Keep only relevant data in the contract state.
+  For example, if you are interested in historical data for all the bids, but your contract logic requires only the highest bid to make a decision, consider :ref:`logging events <contract-instances-logging-events>` instead.
+  An off-chain part of your dApp can then use logs to obtain the historical data.
```
Comment (Bargsteen) [inline source/mainnet/smart-contracts/best-practices/development.rst:170]: Nicely explained 👌

File: source/mainnet/smart-contracts/best-practices/development.rst [hunk 20]
```diff
@@ -0,0 +1,295 @@
...
+.. _best-practices-code-documentation:
+
+Code documentation
+------------------
+
+- Write an outline of the smart contract functionality in the beginning of the file; if the contract implements some standards, mention it.
+- Document decisions/choices in the code.
+- Document entrypoints:
+
+  - What functionality the entrypoint implements?
```
Comment (Bargsteen) [inline source/mainnet/smart-contracts/best-practices/development.rst:259]:
```suggestion
  - What functionality does the entrypoint implement?
```

File: source/mainnet/smart-contracts/best-practices/index.rst (hunk 2 above)
Comment (Bargsteen) [reply source/mainnet/smart-contracts/best-practices/index.rst:7]: Ok, let's keep the structure as is then.

**Review by Bargsteen — APPROVED** [review]
Great work 👍

File: source/mainnet/smart-contracts/best-practices/development.rst (hunk 15 above)
Comment (dg-concordium) [reply source/mainnet/smart-contracts/best-practices/development.rst:30]: I believe an example of pausing also addresses another issue in the documentation backlog: https://github.com/Concordium/concordium.github.io/issues/667

**Review by dg-concordium — APPROVED** [review]
I made a few minor grammar and readability changes.

File: source/mainnet/smart-contracts/best-practices/development.rst (hunk 15 above)
Comment (Bargsteen) [reply source/mainnet/smart-contracts/best-practices/development.rst:30]: No, that is a different thing enterily. The `interrupt`/`resume` events show up when a contract is invoking certain host functions. For example when it invokes another contract. When that happens, the caller is interrupted, then the callee is invoked (which may also produce events), and finally the caller is `resumed`.

File: source/mainnet/smart-contracts/best-practices/development.rst (hunk 19 above)
Comment (annenkov) [reply source/mainnet/smart-contracts/best-practices/development.rst:170]: Thanks :)

File: source/mainnet/smart-contracts/best-practices/development.rst (hunk 15 above)
Comment (annenkov) [reply source/mainnet/smart-contracts/best-practices/development.rst]: I've added a code snippet and pointed to the wCCD example

File: source/mainnet/smart-contracts/best-practices/development.rst [hunk 21]
```diff
@@ -0,0 +1,336 @@
...
+
+A smart contract specification reflects developers' intentions regarding the smart contract functionality.
+It can serve as a guide for the implementation as well as for testing, verifying, and auditing the code.
+Start with an outline of the functionality and gradually refine it into specifications of contract entrypoints.
+
+For the entrypoints, consider the following:
+
+- What is the interface (entrypoints and their parameters)? For example, ``transfer`` takes three parameters: ``from``, ``to`` addresses, and ``amount``.
+- Who can access the entrypoints? For example, the transaction sender must be an operator of the ``from`` account.
+- Which properties of these entrypoints should be satisfied? For example, a successful call of ``transfer(from, to, amount)`` decreases the ``from`` balance and the ``to`` balance by ``amount``.
```
Comment (DOBEN) [inline source/mainnet/smart-contracts/best-practices/development.rst:52]: the ``from`` balance and increases the ``to`` balance by ``amount``.

File: source/mainnet/smart-contracts/best-practices/development.rst [hunk 22]
```diff
@@ -0,0 +1,336 @@
...
+This section provides recommendations for developing smart contracts in Rust.
+See :ref:`introduction` for basic information.
+
+
+.. _best-practices-code-structure:
+
+Recommended structure
+---------------------
+
+- Use ``cargo init`` with an appropriate template to start a new project.
```
Comment (DOBEN) [inline source/mainnet/smart-contracts/best-practices/development.rst:77]: ``cargo concordium init``

File: source/mainnet/smart-contracts/best-practices/development.rst [hunk 23]
```diff
@@ -0,0 +1,336 @@
...
+        mutable
+    )]
+    fn contract_do_something<S: HasStateApi>(
+        ctx: &impl HasReceiveContext,
+        host: &mut impl HasHost<State<S>, StateApiType = S>,
+    ) -> ReceiveResult<()> {
+        // Parse parameters
+        let param: MyParameter = ctx.parameter_cursor().get()?;
+        ...
+        // Perform authorization, potentially using using `ctx` info
```
Comment (DOBEN) [inline source/mainnet/smart-contracts/best-practices/development.rst:107]: `using` is there twice.

**Review by DOBEN — COMMENTED** [review]
A bit late but here are some things that I observed:

**Pull request merged**